applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/BaseHandler.cs

```diff
@@ -0,0 +1,41 @@
+﻿// Copyright(c) 2018 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+using System.Threading.Tasks;
+
+namespace GoogleHomeAspNetCoreDemoServer.Dialogflow.Intents
+{
+    /// <summary>
+    /// Base class for intent handlers. 
+    /// </summary>
+    public abstract class BaseHandler
+    {
+        protected readonly Conversation _conversation;
+
+        /// <summary>
+        /// Initializes class with the conversation.
+        /// </summary>
+        /// <param name="conversation">Conversation</param>
+        public BaseHandler(Conversation conversation)
+        {
+            _conversation = conversation;
+        }
+
+        /// <summary>
+        /// Handles the conversation request.
+        /// </summary>
+        /// <param name="req"></param>
+        /// <returns></returns>
+        public abstract Task<object> Handle(ConvRequest req);
```

Comment:
I would strongly suggest changing how this works. There are a few problems currently:
* All handlers are forced to be asynchronous, as they all have to conform to this signature that returns a `Task<...>`.
* The `switch (result)` block in the `Handle` method in the `Conversation` class is pointless as it stands, as all returned values will be `Task<string>`.
* This means some handlers are declared as `async`, but don't use `await`; which is not a good pattern.
* `Task<string>` is the only actual type that must be returned; `Task<object>` is broader than required.
* (And this method should really be called `HandleAsync`).

One way to improve this (in my opinion!) would be to declare two handlers, one sync and one async:
```C#
public virtual Task<string> HandleAsync(ConvRequest req) => return Task.FromResult<string>(null);
public virtual string Handle(ConvRequest req) => return null;
```

Deriving classes only overrides one of these methods, depending on whether they want to do a sync or async implementation.

Then change the code in `Conversation.Handle[Async]` to be more like:
```C#
return handler.Handle(req) ?? await handler.HandleAsync(req) ??
    DialogflowApp.Tell("Error. Handler did not return a valid response.")
```
And the switch statement is no longer required.

Reply:
Although I wasn't happy about having two methods for essentially the same thing, I think there's no better alternative. So, I went ahead and implemented what you suggested. Thanks!